### PR TITLE
Settings actions reset to default

### PR DESCRIPTION
**Pull request by nnngrach**
https://github.com/osmandapp/OsmAnd-Issues/issues/93

### CODE REVIEW

File: Sources/Helpers/OAAppSettings.m [hunk 1]
```diff
@@ -1855,6 +1855,7 @@ - (instancetype) init
 
         _settingShowMapRulet = [[NSUserDefaults standardUserDefaults] objectForKey:settingShowMapRuletKey] ? [[NSUserDefaults standardUserDefaults] boolForKey:settingShowMapRuletKey] : YES;
         _settingAppMode = [OAProfileInteger withKey:settingAppModeKey defValue:0];
+        [_registeredPreferences setObject:_settingAppMode forKey:settingAppModeKey];
```
Comment (max-klaus) [inline Sources/Helpers/OAAppSettings.m:1858]: **forKey:** must be equal to android setting key!

File: Sources/Helpers/OAAppSettings.m [hunk 2]
```diff
@@ -2279,6 +2292,37 @@ - (OAProfileSetting *) getSettingById:(NSString *)stringId
...
+    
+    if (!appMode.isCustomProfile)
+    {
+        [self.userProfileName resetModeToDefault:appMode];
+        [self.profileIconName resetModeToDefault:appMode];
+        [self.profileIconColor resetModeToDefault:appMode];
+    }
+    
+    //proflie undepended settings
+    [self hideGpx:_mapSettingVisibleGpx];
```
Comment (max-klaus) [inline Sources/Helpers/OAAppSettings.m:2320]: profile undepended should not be reset

File: Sources/OAApplicationMode.m [hunk 3]
```diff
@@ -601,12 +601,48 @@ + (void) saveProfile:(OAApplicationMode *)appMode
...
+    NSString *backupFolderPath = [[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject] stringByAppendingPathComponent:@"osfBackup"];
+    
+    BOOL isDir = YES;
+    if (![fileManager fileExistsAtPath:backupFolderPath isDirectory:&isDir])
+          [fileManager createDirectoryAtPath:backupFolderPath withIntermediateDirectories:YES attributes:nil error:nil];
+    
+    if (appMode.stringKey)
+    {
+        NSString *backupFilePath = [[backupFolderPath stringByAppendingPathComponent:appMode.stringKey] stringByAppendingPathExtension:@"plst"];
+        [appMode.toJson writeToFile:backupFilePath atomically:YES];
```
Comment (max-klaus) [inline Sources/OAApplicationMode.m:623]: Not only appMode.toJson should be saved, but all profile depended settings! Use OASettingsExporter.

File: Sources/Data/OAMapStyleSettings.mm [hunk 4]
```diff
@@ -353,4 +353,24 @@ -(void) save:(OAMapStyleParameter *)parameter refreshMap:(BOOL)refreshMap
...
+    NSArray<OAMapStyleParameter *> *clearParameters = [NSArray arrayWithArray:self.parameters];
+    
+    dispatch_async(dispatch_get_main_queue(), ^{
+        for (OAMapStyleParameter *p in clearParameters)
+        {
+            p.value = p.defaultValue;
+            p.storedValue = p.defaultValue;
+            p.mapPresetName = appModeName;
+
+            NSString *name = [NSString stringWithFormat:@"%@_type_%@_%@", p.mapStyleName, p.mapPresetName, p.name];
```
Comment (max-klaus) [inline Sources/Data/OAMapStyleSettings.mm:367]: What is **%@_type_%@_%@** ? Why not use method **save:** ? Copy - paste?

File: Sources/Data/OAMapStyleSettings.mm [hunk 5]
```diff
@@ -353,4 +353,24 @@ -(void) save:(OAMapStyleParameter *)parameter refreshMap:(BOOL)refreshMap
         [[[OsmAndApp instance] mapSettingsChangeObservable] notifyEvent];
 }
 
+-(void) resetMapStyleForAppMode:(NSString *)appModeName
+{
+    NSArray<OAMapStyleParameter *> *clearParameters = [NSArray arrayWithArray:self.parameters];
```
Comment (max-klaus) [inline Sources/Data/OAMapStyleSettings.mm:358]: What if **self.parameters** loaded with differ appModeName?

File: Sources/Helpers/OAAppSettings.m [hunk 6]
```diff
@@ -1854,7 +1854,8 @@ - (instancetype) init
         _settingMapLanguageTranslit = [[NSUserDefaults standardUserDefaults] objectForKey:settingMapLanguageTranslitKey] ? [[NSUserDefaults standardUserDefaults] boolForKey:settingMapLanguageTranslitKey] : NO;
 
         _settingShowMapRulet = [[NSUserDefaults standardUserDefaults] objectForKey:settingShowMapRuletKey] ? [[NSUserDefaults standardUserDefaults] boolForKey:settingShowMapRuletKey] : YES;
-        _settingAppMode = [OAProfileInteger withKey:settingAppModeKey defValue:0];
+        _dayNightMode = [OAProfileInteger withKey:settingDayNightModeKey defValue:0];
+        [_registeredPreferences setObject:_dayNightMode forKey:settingDayNightModeKey];
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:1858]: Keys should be identical to Android. You cannot use our internal settings' keys. All keys can be found in the profile .json file inside of .osf.

File: Sources/Helpers/OAAppSettings.h [hunk 7]
```diff
@@ -11,9 +11,11 @@
 #import "OAApplicationMode.h"
 
 #define kNotificationSetProfileSetting @"kNotificationSetProfileSetting"
+#define kUpdateWidgestsVisibilityNotification @"kUpdateWidgestsVisibilityNotification"
+#define kResetWidgetsSettingsNotification @"kResetWidgetsSettingsNotification"
 #define VOICE_PROVIDER_NOT_USE @"VOICE_PROVIDER_NOT_USE"
 
-#define settingAppModeKey @"settingAppModeKey"
+#define daynight_mode @"daynight_mode"
```
Comment (max-klaus) [inline Sources/Helpers/OAAppSettings.h:18]: #define settingAppModeKey @"settingAppModeKey" > #define settingAppModeKey @"settingAppModeKey"

File: Sources/Helpers/OASettingsHelper.mm [hunk 8]
```diff
@@ -520,8 +527,31 @@ - (BOOL) readFromFile:(NSString *)filePath error:(NSError * _Nullable *)error
     [self.item applyRoutingPreferences:routingSettings];
     
     [OsmAndApp.instance.data.mapLayerChangeObservable notifyEvent];
+}
+
+- (void) saveToBackup:(NSDictionary<NSString *, NSString *> *)settings withFilename:(NSString *)filename
```
Comment (max-klaus) [inline Sources/Helpers/OASettingsHelper.mm:532]: It is not clear why applyReadedSettings, saveToBackup and restoreFromBackup added here. Looks like wrong approach.

File: Sources/Helpers/OAMapStyleTitles.m [hunk 9]
```diff
@@ -51,4 +51,20 @@ + (int) getSortIndexForTitle:(NSString *)title
         return 7;
 }
 
++ (NSArray<NSString *> *)getMapStyleRenderKeys
```
Comment (max-klaus) [inline Sources/Helpers/OAMapStyleTitles.m:54]: Remove hardcode. See _offlineMapSources

File: Sources/Controllers/2.0/BottomSheet/PluginResetBottomSheet/OAPluginResetBottomSheetViewController.m [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+{
+    return 16.0;
+}
+
+- (void)tableView:(UITableView *)tableView willDisplayFooterView:(UIView *)view forSection:(NSInteger)section
+{
+    view.hidden = YES;
+}
+
+@synthesize vwController;
```
Comment (max-klaus) [inline Sources/Controllers/2.0/BottomSheet/PluginResetBottomSheet/OAPluginResetBottomSheetViewController.m:291]: ?

File: Sources/Helpers/OAAppSettings.m [hunk 11]
```diff
@@ -3026,6 +3021,55 @@ - (NSString *) getDefaultVoiceProvider
...
 
+- (BOOL) getOverlayOpacitySliderVisibility
+{
+    return [_layerTransparencySeekbarMode get] == LAYER_TRANSPARENCY_SEEKBAR_MODE_OVERLAY || [_layerTransparencySeekbarMode get] == LAYER_TRANSPARENCY_SEEKBAR_MODE_ALL;
+}
+
+- (void) setOverlayOpacitySliderVisibility:(BOOL)isVisible
+{
+    if (isVisible)
+        if ([_layerTransparencySeekbarMode get] == LAYER_TRANSPARENCY_SEEKBAR_MODE_UNDERLAY)
```
Comment (max-klaus) [inline Sources/Helpers/OAAppSettings.m:3028]: what if _layerTransparencySeekbarMode == ALL?

File: Sources/Helpers/OAAppSettings.m [hunk 12]
```diff
@@ -3026,6 +3021,55 @@ - (NSString *) getDefaultVoiceProvider
...
+        else
+            [_layerTransparencySeekbarMode set:LAYER_TRANSPARENCY_SEEKBAR_MODE_OVERLAY];
+   else
+        if ([_layerTransparencySeekbarMode get] == LAYER_TRANSPARENCY_SEEKBAR_MODE_ALL)
+            [_layerTransparencySeekbarMode set:LAYER_TRANSPARENCY_SEEKBAR_MODE_OVERLAY];
+        else
+            [_layerTransparencySeekbarMode set:LAYER_TRANSPARENCY_SEEKBAR_MODE_OFF];
+}
+
+- (OAProfileBoolean *) getCustomRoutingBooleanProperty1:(NSString *)attrName defaultValue:(BOOL)defaultValue
```
Comment (max-klaus) [inline Sources/Helpers/OAAppSettings.m]: 1?

File: Sources/Helpers/OAAppSettings.h [hunk 13]
```diff
@@ -421,9 +427,12 @@ typedef NS_ENUM(NSInteger, EOARulerWidgetMode)
 @property (nonatomic) OAProfileBoolean *mapSettingShowFavorites;
 @property (nonatomic) OAProfileBoolean *mapSettingShowOfflineEdits;
 @property (nonatomic) OAProfileBoolean *mapSettingShowOnlineNotes;
-@property (nonatomic) OAProfileBoolean *mapSettingShowOverlayOpacitySlider;
-@property (nonatomic) OAProfileBoolean *mapSettingShowUnderlayOpacitySlider;
 @property (nonatomic) NSArray *mapSettingVisibleGpx;
+@property (nonatomic) OAProfileInteger *layerTransparencySeekbarMode; // 0 - overlay, 1 - underlay, 2 - off, 3 - undefined, 4 - overlay&underlay
+- (BOOL) getOverlayOpacitySliderVisibility;
+- (BOOL) getUnderlayOpacitySliderVisibility;
+- (void) setOverlayOpacitySliderVisibility:(BOOL)isVisible;
```
Comment (max-klaus) [inline Sources/Helpers/OAAppSettings.h:434]: Visibility:(BOOL)isVisible; > Visibility:(BOOL)visibility;

File: Sources/Controllers/2.0/OAMapSettingsMainScreen.mm [hunk 14]
```diff
@@ -232,11 +232,11 @@ - (void) setupView
...
         NSMutableArray *categoriesList = [NSMutableArray array];
         NSString *modeStr;
-        if ([_settings.settingAppMode get] == APPEARANCE_MODE_DAY)
+        if ([_settings.dayNightMode get] == APPEARANCE_MODE_DAY)
             modeStr = OALocalizedString(@"map_settings_day");
-        else if ([_settings.settingAppMode get] == APPEARANCE_MODE_NIGHT)
+        else if ([_settings.dayNightMode get] == APPEARANCE_MODE_NIGHT)
             modeStr = OALocalizedString(@"map_settings_night");
-        else if ([_settings.settingAppMode get] == APPEARANCE_MODE_AUTO)
+        else if ([_settings.dayNightMode get] == APPEARANCE_MODE_AUTO)
```
Comment (max-klaus) [inline Sources/Controllers/2.0/OAMapSettingsMainScreen.mm:239]: if constants start with APPEARANCE, settings should start with appearance as well. I.e. **appearanceMode**

File: Sources/Helpers/OAProfileSettingsResetHelper.h [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+//  OAProfileSettingsResetHelper.h
+//  OsmAnd
+//
+//  Created by nnngrach on 08.09.2020.
+//  Copyright © 2020 OsmAnd. All rights reserved.
+//
+
+#import "OASettingsHelper.h"
+
+@interface OAProfileSettingsResetHelper : NSObject
```
Comment (max-klaus) [inline Sources/Helpers/OAProfileSettingsResetHelper.h:11]: Not clear why do we need such helper class. @PaulStets please review it and maybe compare with android code.

File: Sources/Controllers/2.0/BottomSheet/PluginResetBottomSheet/OAPluginResetBottomSheetViewController.m [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+    NSString *type = item[@"type"];
+    if ([type isEqualToString:@"OABottomSheetHeaderDescrButtonCell"])
+    {
+        return UITableViewAutomaticDimension;
+    }
+    else if ([type isEqualToString:@"OADescrTitleCell"])
+    {
+        return [OADescrTitleCell getHeight:item[@"title"] desc:item[@"description"] cellWidth:DeviceScreenWidth];
+    }
+    else if ([type isEqualToString:@"OASettingSwitchNoImageCell"])
```
Comment (PaulStets) [inline Sources/Controllers/2.0/BottomSheet/PluginResetBottomSheet/OAPluginResetBottomSheetViewController.m:133]: Are all of these cell types are used on this screen? (Same for cellForRow:)

File: Sources/Controllers/2.0/Settings/OAConfigureProfileViewController.m [hunk 17]
```diff
@@ -156,12 +159,14 @@ - (void) generateData
         @"key" : @"reset_to_default"
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/OAConfigureProfileViewController.m:159]: Missing a check (see ConfigureProfileFragment.java :341 ```private void setupResetToDefaultPref()```)

File: Sources/Helpers/OAProfileSettingsResetHelper.h (hunk 15 above)
Comment (PaulStets) [reply Sources/Helpers/OAProfileSettingsResetHelper.h:11]: This class is indeed redundant, all the reset logic in Android is in ConfigureProfileFragment.java: 171 ```public void resetAppModePrefs(ApplicationMode appMode)```. The only difference is: unlike in Android all  the Rendering and Routing settings should be reset separately (see OASettingsHelper.mm import logic).

File: Sources/Helpers/OAProfileSettingsResetHelper.m [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+    [OsmAndApp.instance.data.mapLayerChangeObservable notifyEvent];
+}
+
++ (void) saveToBackup:(NSDictionary<NSString *, NSString *> *)settings withFilename:(NSString *)filename
+{
+    NSFileManager *fileManager = NSFileManager.defaultManager;
+    NSString *backupFolderPath = [[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject] stringByAppendingPathComponent:@"osfBackup"];
+    
+    BOOL isDir = YES;
+    if (![fileManager fileExistsAtPath:backupFolderPath isDirectory:&isDir])
```
Comment (PaulStets) [inline Sources/Helpers/OAProfileSettingsResetHelper.m:75]: Note

Attempting to predicate behavior based on the current state of the file system or a particular file on the file system is not recommended. Doing so can cause odd behavior or race conditions. It’s far better to attempt an operation (such as loading a file or creating a directory), check for errors, and handle those errors gracefully than it is to try to figure out ahead of time whether the operation will succeed.  More: https://developer.apple.com/documentation/foundation/nsfilemanager/1415645-fileexistsatpath?language=objc

File: Sources/OAApplicationMode.h [hunk 19]
```diff
@@ -62,6 +62,8 @@
 - (instancetype)initWithName:(NSString *)name stringKey:(NSString *)stringKey;
 
 - (NSDictionary *) toJson;
++ (void) saveBackupOfAppMode:(OAApplicationMode *)appMode;
```
Comment (PaulStets) [inline Sources/OAApplicationMode.h:65]: This is an incorrect place for this logic. Settings should be collected from backup using the OASettingsHelper (again ConfigureProfileFragment.java: 171 public void resetAppModePrefs(ApplicationMode appMode))

File: Sources/Controllers/2.0/BottomSheet/OABottomSheetTwoButtonsViewController.mm [hunk 20]
```diff
@@ -43,28 +43,40 @@ - (void)additionalSetup
 
 - (void) setupButtons
 {
+    [self layoutButtons];
     if (_doneButton.hidden)
```
Comment (PaulStets) [inline Sources/Controllers/2.0/BottomSheet/OABottomSheetTwoButtonsViewController.mm:47]: Why do you need to hide the button if it's hidden? Also, why do you call [self layoutButtons] twice (here and in [self hideDoneButton];)? It's better to call it after [self hideDoneButton] and remove it from the  hideDoneButton method.

File: Sources/Helpers/OASettingsHelper.h [hunk 21]
```diff
@@ -133,6 +142,8 @@ typedef enum : NSUInteger {
 
 @interface OASettingsItemJsonReader : OASettingsItemReader<OASettingsItem *>
 
+- (void) applyReadedSettings:(NSDictionary<NSString *, NSString *> *)settings;
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.h:145]: applyReadSettings or applyParsedSettings (readed doesn't exsist in English).

File: Sources/Helpers/OASettingsHelper.mm [hunk 22]
```diff
@@ -208,6 +209,130 @@ - (void) onDuplicatesChecked:(NSArray<OASettingsItem *>*)duplicates items:(NSArr
...
+            [self updateCopiedOrResetPrefs:appMode];
+        }
+    }
+    return appMode;
+}
+
+- (void) updateCopiedOrResetPrefs:(OAApplicationMode *)appMode
+{
+    if ([OAAppSettings sharedManager].applicationMode == appMode)
+        [[NSNotificationCenter defaultCenter] postNotificationName:kUpdateWidgestsVisibilityNotification object:nil userInfo:nil];
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:241]: We prefer using ```OAAutoObserverProxy```

File: Sources/Helpers/OASettingsHelper.mm [hunk 23]
```diff
@@ -208,6 +209,130 @@ - (void) onDuplicatesChecked:(NSArray<OASettingsItem *>*)duplicates items:(NSArr
...
+    
+    [[[OsmAndApp instance] mapSettingsChangeObservable] notifyEvent];
+}
+
+- (void) resetPreferencesForProfile:(OAApplicationMode *)appMode
+{
+    OAAppSettings.sharedManager.applicationMode = appMode;
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:248]: Why do you need to set the profile you are resetting as an active profile?

File: Sources/Helpers/OASettingsHelper.mm [hunk 24]
```diff
@@ -208,6 +209,130 @@ - (void) onDuplicatesChecked:(NSArray<OASettingsItem *>*)duplicates items:(NSArr
...
+    [OAAppSettings.sharedManager resetAllProfileSettingsForMode:appMode];
+    [OAAppData.defaults resetProfileSettingsForMode:appMode];
+    [OAMapStyleSettings.sharedInstance resetMapStyleForAppMode:appMode.variantKey];
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:251]: MapStyleSettings need to be created for this specific appMode.

File: Sources/Helpers/OASettingsHelper.mm [hunk 25]
```diff
@@ -208,6 +209,130 @@ - (void) onDuplicatesChecked:(NSArray<OASettingsItem *>*)duplicates items:(NSArr
...
+    [fileManager createDirectoryAtPath:backupFolderPath withIntermediateDirectories:YES attributes:nil error:nil];
+    
+    NSString *backupFilePath = [[backupFolderPath stringByAppendingPathComponent:filename] stringByAppendingPathExtension:@"plst"];
+    [settings writeToFile:backupFilePath atomically:YES];
+}
+
+- (void) saveToBackupCustomProfileAppMode:(OAApplicationMode *)appMode
+{
+    NSFileManager *fileManager = NSFileManager.defaultManager;
+    NSString *backupFolderPath = [[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) firstObject] stringByAppendingPathComponent:@"osfBackup"];
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:298]: We have all the root paths stored in OsmAndAppImpl .

File: Sources/Helpers/OASettingsHelper.mm [hunk 26]
```diff
@@ -208,6 +209,130 @@ - (void) onDuplicatesChecked:(NSArray<OASettingsItem *>*)duplicates items:(NSArr
...
+    [fileManager createDirectoryAtPath:backupFolderPath withIntermediateDirectories:YES attributes:nil error:nil];
+    
+    if (appMode.stringKey)
+    {
+        NSString *backupFilePath = [[backupFolderPath stringByAppendingPathComponent:appMode.stringKey] stringByAppendingPathExtension:@"plst"];
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:303]: Why ```plst```? See getBackupFileForCustomMode() in ConfigureProfileFragment.java

Comment (PaulStets) [reply Sources/Helpers/OASettingsHelper.mm:303]: Also, see ProfileAppearanceFragment.java: 818 ```private void saveProfileBackup(ApplicationMode mode)```

File: Resources/en.lproj/Localizable.strings [hunk 27]
```diff
@@ -765,7 +765,10 @@
 "selected_profile" = "Selected profile";
 "app_profiles" = "App profiles";
 "profile_sett_descr" = "Affect selected application profile";
-"export_profile_descr" = "To import a profile, tap on the needed file (*.osf) in any application and select \"Copy to OsmAnd Maps\"";
+"export_profile_descr" = "You can export all profile settings. Copy all settings from other profile. Reset all profile settings to default state.";
```
Comment (PaulStets) [inline Resources/en.lproj/Localizable.strings:768]: Why did you delete the original string? Please re-add the deleted string with a different id and use it in OAMainSettingsView controller. As for the new string, please correct it to proper English: "You can export all profile settings, copy all settings from another profile or reset all profile settings to default state."

File: Resources/en.lproj/Localizable.strings [hunk 28]
```diff
@@ -765,7 +765,10 @@
...
+
+"reset_to_default" = "Reset to default";
+"reset_profile_action_descr" = "Reset all profile settings to installation defaults. Tapping Reset will disvard all your changes, this action can't be undo.";
```
Comment (PaulStets) [inline Resources/en.lproj/Localizable.strings:771]: "will discard", "can't be undone"

File: Sources/Controllers/2.0/BottomSheet/PluginResetBottomSheet/OAPluginResetBottomSheetViewController.m [hunk 29]
```diff
@@ -0,0 +1,341 @@
...
+- (void)applyLocalization
+{
+    [self.cancelButton setTitle:OALocalizedString(@"shared_string_cancel") forState:UIControlStateNormal];
+    [self.doneButton setTitle:OALocalizedString(@"shared_string_reset") forState:UIControlStateNormal];
+}
+
+-(void) doneButtonPressed:(id)sender
+{
+    if (self.delegate)
+        [self.delegate onAppModeChangedByPluginResetBottomSheet];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/BottomSheet/PluginResetBottomSheet/OAPluginResetBottomSheetViewController.m:336]: onPluginSettingsReset

File: Sources/Helpers/OAAppSettings.m [hunk 30]
```diff
@@ -2301,6 +2309,31 @@ - (OAProfileSetting *) getSettingById:(NSString *)stringId
     return [_registeredPreferences objectForKey:stringId];
 }
 
+- (void) resetAllProfileSettingsForMode:(OAApplicationMode *)appMode
+{
+    for (OAProfileSetting *value in [_registeredPreferences objectEnumerator].allObjects)
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:2314]: Aren't all profile settings inherit from OAProfileSetting class? Also, what about other types of ProfileSetting?

File: Sources/Helpers/OAAppSettings.m [hunk 31]
```diff
@@ -2334,7 +2367,7 @@ - (void) setSettingMapLanguageTranslit:(BOOL)settingMapLanguageTranslit
 
 - (void) setAppMode:(int)settingAppMode
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:2368]: It would be better to also rename this method.

File: Sources/Helpers/OASettingsHelper.mm [hunk 32]
```diff
@@ -208,6 +214,135 @@ - (void) onDuplicatesChecked:(NSArray<OASettingsItem *>*)duplicates items:(NSArr
...
+{
+    NSString *fileName = appMode.stringKey;
+    NSString *backupDir = [[OsmAndApp instance].documentsPath stringByAppendingPathComponent:BACKUP_INDEX_DIR];
+    [self createDirectoryIfNotExist:backupDir];
+    
+    NSString *backupFilePath = [[backupDir stringByAppendingPathComponent:fileName] stringByAppendingPathExtension:OSMAND_SETTINGS_FILE_EXT];
+    [appMode.toJson writeToFile:backupFilePath atomically:YES];
+}
+
+- (NSDictionary<NSString *, NSString *> *) getBackupFileForCustomProfile:(OAApplicationMode *)appMode
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:309]: What's the difference between this and the next method?

File: Sources/Helpers/OAAppSettings.m (hunk 30 above)
Comment (nnngrach) [reply Sources/Helpers/OAAppSettings.m:2314]: mark as resolved

File: Sources/Controllers/2.0/OATargetInfoViewController.mm [hunk 33]
```diff
@@ -626,15 +626,17 @@ - (void) removeDublicatesFromCards:(NSMutableArray<OAAbstractCard *> *)cards
 {
     NSMutableArray *wikimediaCards = [NSMutableArray new];
     NSMutableArray *mapilaryCards = [NSMutableArray new];
+    OAMapillaryContributeCard *mapilaryConributeCard = nil;
```
Comment (max-klaus) [inline Sources/Controllers/2.0/OATargetInfoViewController.mm:629]: mapilaryConributeCard  > mapilaryContributeCard

File: Sources/Controllers/Map/Widgets/OAMapWidgetRegistry.mm [hunk 34]
```diff
@@ -36,22 +38,8 @@ - (instancetype) init
...
-            else
-            {
-                NSMutableSet<NSString *> *set = [NSMutableSet set];
-                [_visibleElementsFromSettings setObject:set forKey:ms];
-                NSArray<NSString *> *split = [mpf componentsSeparatedByString:SETTINGS_SEPARATOR];
-                [set addObjectsFromArray:split];
-            }
-        }
+        [self loadVisibleElementsFromSettings];
+        _widgetSettingsChangeDoneObserver = [[OAAutoObserverProxy alloc] initWith:self withHandler:@selector(onWidgetSettingsReset:withKey:) andObserve:[OsmAndApp instance].widgetSettingsChangingStartObservable];
```
Comment (max-klaus) [inline Sources/Controllers/Map/Widgets/OAMapWidgetRegistry.mm:42]: Names mismatch _widgetSettingsChangeDoneObserver and widgetSettingsChangingStartObservable. Looks like widgetSettingResetObservable should be used in both cases.

File: Sources/Data/OAMapStyleSettings.h [hunk 35]
```diff
@@ -66,4 +64,6 @@ typedef NS_ENUM(NSInteger, OAMapStyleValueDataType)
 - (void) save:(OAMapStyleParameter *)parameter;
 - (void) save:(OAMapStyleParameter *)parameter refreshMap:(BOOL)refreshMap;
 
+-(void) resetMapStyleForAppMode:(NSString *)appModeName;
```
Comment (max-klaus) [inline Sources/Data/OAMapStyleSettings.h:67]: appModeName is mapPresetName for OAMapStyleSettings. Thus there should not be confusion with names in one class. Looks like method should be: 
`-(void) resetMapStyleForPresetName:(NSString *)mapPresetName;`

File: Sources/Helpers/OAAppSettings.m [hunk 36]
```diff
@@ -1965,6 +1963,7 @@ - (instancetype) init
         _customAppModes = [NSUserDefaults.standardUserDefaults objectForKey:customAppModesKey] ? [NSUserDefaults.standardUserDefaults stringForKey:customAppModesKey] : @"";
 
         _mapInfoControls = [OAProfileString withKey:mapInfoControlsKey defValue:@""];
+        [_registeredPreferences setObject:_mapInfoControls forKey:mapInfoControlsKey];
```
Comment (max-klaus) [inline Sources/Helpers/OAAppSettings.m:1966]: Are these keys synchronized with android?

File: Sources/Helpers/OAAppSettings.m [hunk 37]
```diff
@@ -27,10 +27,11 @@
 #define settingEnable3DViewKey @"settingEnable3DView"
 #define settingDoNotShowPromotionsKey @"settingDoNotShowPromotionsKey"
 #define settingUseFirebaseKey @"settingUseFirebaseKey"
-#define settingExternalInputDeviceKey @"settingExternalInputDeviceKey"
+#define settinlastMapSourcegExternalInputDeviceKey @"settingExternalInputDeviceKey"
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:30]: Typo?

File: Sources/Helpers/OAAppSettings.m [hunk 38]
```diff
@@ -27,10 +27,11 @@
...
 #define metricSystemChangedManuallyKey @"metricSystemChangedManuallyKey"
 #define liveUpdatesPurchasedKey @"liveUpdatesPurchasedKey"
 #define settingOsmAndLiveEnabledKey @"settingOsmAndLiveEnabledKey"
+#define settingExternalInputDeviceKey @"settingExternalInputDeviceKey"
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:34]: Duplicate key?

File: Sources/Helpers/OAAppSettings.m [hunk 39]
```diff
@@ -136,6 +137,7 @@
 #define angularUnitsKey @"angularUnits"
 #define speedLimitExceedKey @"speedLimitExceed"
 #define switchMapDirectionToCompassKey @"switchMapDirectionToCompass"
+#define wakeOnVoiceIntKey @"wakeOnVoiceInt"
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:140]: This setting was removed yesterday, no need to re add it

File: Sources/Helpers/OAAppSettings.m [hunk 40]
```diff
@@ -2108,6 +2109,12 @@ - (instancetype) init
         _routeRecalculationDistance = [OAProfileDouble withKey:routeRecalculationDistanceKey defValue:0.];
         [_registeredPreferences setObject:_routeRecalculationDistance forKey:@"routing_recalc_distance"];
 
+        _wakeOnVoiceInt = [OAProfileInteger withKey:wakeOnVoiceIntKey defValue:0];
```
Comment (PaulStets) [inline Sources/Helpers/OAAppSettings.m:2112]: This was deleted yesterday

File: Sources/Helpers/OASettingsHelper.mm [hunk 41]
```diff
@@ -178,10 +188,6 @@ - (void) onSettingsCollectFinished:(BOOL)succeed empty:(BOOL)empty items:(NSArra
             else if (item.pluginId.length == 0)
                 [pluginIndependentItems addObject:item];
         }
-//        for (OAPluginSettingsItem *pluginItem in pluginSettingsItems)
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:181]: This code was commented out to implement it later

File: Sources/Helpers/OASettingsHelper.mm [hunk 42]
```diff
@@ -650,6 +670,11 @@ - (BOOL)exists
 }
 
 - (void)readFromJson:(id)json error:(NSError * _Nullable __autoreleasing *)error
+{
+    [self readFromJsonWithoutBackup:json error:error];
+}
+
+- (void)readFromJsonWithoutBackup:(id)json error:(NSError * _Nullable __autoreleasing *)error
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:677]: No such method in Android. It should be called readFromJson

File: Sources/Helpers/OAAppSettings.m (hunk 38 above)
Comment (nnngrach) [reply Sources/Helpers/OAAppSettings.m:33]: Deleted the previous one.
